### ISSUE TITLE
Problem assigning topics do training texts

### ISSUE THREAD

**Issue by pvcastro** [MaartenGr/BERTopic#249]
Hi there!

I'm facing a strange problem here. I trained a topic model at a robust server, using around 300k tweet messages, using the following parameters:

```
topic_model = BERTopic(low_memory=True, nr_topics=number_of_topics, verbose=True, n_gram_range=(1, 3),
                           vectorizer_model=CountVectorizer(ngram_range=(1, 3), stop_words=stopwords))
topic_model.fit_transform(tweets)
topic_model.save(output_path)
```

Since it's a large dataset and I'm using trigrams, it consumes a lot of RAM, so that's why I ran on a large server.
Then I download the model to my local machine to check the results. Here's what I'm doing:

```
topic_model = BERTopic.load(path=default_3gram_5t)
topic_model.get_topic_info()
```
These are the results (trained with 5 topics):

![bertopic_topics](https://user-images.githubusercontent.com/12713359/134553998-79ae8439-9279-4a4f-8455-5d4e1a64e63f.png)

However, when I run transform for the **same** documents to get their assigned topics, using this:

```
topics, probs = topic_model.transform(tweets)
```

I get a very different distribution from the one above. Almost all of them are assigned to the -1 topic, and only around 400 are distributed among topics 0 to 3. No tweet gets assigned to topic 4. 

Can you tell me what might be happening? 
I'm using 0.9.1 on both server and local machines.

Thanks!

**Comment by MaartenGr** [comment]
I believe that there are two things happening here. The first you are mentioning is the RAM issue. I am quite sure that could quickly be resolved by following the FAQ regarding the memory issue [here](https://maartengr.github.io/BERTopic/faq.html#i-am-facing-memory-issues-help). I would advise focusing especially on making sure the `min_df` is set to a minimum of 5. 

Secondly, with 300k tweets, you can be quite certain that they contain a bit more than just 5 topics. Most likely, I would expect to extract a couple of hundred topics from 300k documents. By reducing several hundred topics to only 5 you are losing so much information that it is likely those will end up in the -1 topic. I would suggest setting `nr_topics` to either `"auto"` or to at least 50% of the initially generated number of topics. 

Hopefully, this helps a bit. If not, please let me know and I'll be glad to help!

**Comment by pvcastro** [comment]
Hi @MaartenGr , thanks for the prompt response!
First, I did try those recommendations from your FAQ before moving to the server, but I was getting memory issues nonetheless. You can see that I'm already using `low_memory=True` and calculate_probabilities is with the default False. The only thing missing there is the min_df=5, but I was already getting memory errors using this, and I wasn't even using trigrams yet. 

I understand that I have a lot of tweets for so few topics, but I got this "requirement" from our business team, which provided the curated list of tweets, and told us that there shouldn't be other relevant topics above 15 or 20. I didn't mention here, but I have other 3 versions with these settings, but varying topics from 5 to 20. From their perspective, the extracted topics and keywords were reasonable enough. 

Anyway, do you think that the small number of topics would explain this issue of not assigning the original topics from training?

Thanks!

**Comment by MaartenGr** [comment]
The `min_df` parameter should help when you have large amounts of data as the resulting sparse matrix can be quite big due to having an extensive vocabulary. The main issue is often sorting the c-TF-IDF sparse matrix row-wise for which it has been difficult to find a solution that does not involve adding dependencies. Typically, increasing the `min_df` param should significantly reduce the neccesary RAM. 

When you are forcefully reducing the number of topics, that is merging them together, the resulting topics lose their meaning. If you have 500 initial topics and you merge them together until 5 topics remain, then it is quite likely those 5 topics are too abstract. Instead, you can use `nr_topics="auto"` to reduce the topics by only merging those that are similar. Then, in order to comply with business requirements, manually map the irrelevant topics to the -1 topic. That way, you keep the accuracy of the model whilst preventing topics that are unrelated to be merged into existing topics. 

Yes, I think that because of the explanation above, many topics being forced together, it is difficult for the model to assign a document to a specific topic.

**Comment by pvcastro** [comment]
@MaartenGr thanks, your explanation makes sense to me. 
But now I'm wondering... Isn't a min_df of 5 too much for tweets? Considering that they are small texts, and the same word isn't likely to repeat on them.

**Comment by MaartenGr** [comment]
I understand the hesitancy when setting `min_df`. However, it checks whether a word repeats itself between texts and not within texts. In other words, it does not matter whether a tweet has repeating words but whether a word is found in multiple tweets. For example, if a word can only be found in a specific tweet and not all other 299,999 tweets, then it will be ignored. A word is only interesting if it can be found at least a certain number of times, in this case 5, across **all** tweets.

**Comment by pvcastro** [comment]
I see, that's great then. I'll close this now and retrain the models. In case I can't apply your suggestions I'll reopen it. Thanks!
